### DISCUSSION TITLE
Support for posits and bfloat16

### DISCUSSION THREAD

**Discussion by spytheman** [vlang/v#11654] [Ideas]
Please add posit[1] and bfloat16[2].
Posits/unums are used in HPC computing and bfloat16 on neural networks training. Might be of interest in the future. 
[1] https://en.wikipedia.org/wiki/Unum_(number_format)
      https://posithub.org/index
[2] https://en.wikipedia.org/wiki/Bfloat16_floating-point_format
(extracted from #3961)

**Comment by dumblob** [comment]
Motivating example: https://github.com/stillwater-sc/universal

**Comment by ulises-jeremias** [comment]
@spytheman what do you think about having this on VSL instead of vlib? At least the `bfloat16`, I think will be good to have it on VSL as part of the other stuff we have in there and then we can use it to define other librerias using it as backend. We are already using VSL on vtl, and the plan is to have generic NN support in VTL so maybe there is no need to have `bfloat16` on vlib

Same could apply for `posit`, but maybe it is good to have it on vlib since it's more general purpose

**Comment by JalonSolov** [comment]
Depends.  It would be good to see an implementation in V, so the performance and accuracy could be verified.
If posits really are a "drop-in" replacement for regular floats, and really are faster, it might be a good idea to replace the floating point code in V altogether.

However, other considerations include size differences in both V and user code, as well conversions... will the numbers look the same (or enough the same) such that converting <other language> code to V will work the same?  If the answer is "no, you'll get different results from the same calculations", that's a negative effect - even if the new results are "more correct".

**Reply by ulises-jeremias** [reply]
oh yes, just in case to avoid confusion, I was not meaning to not implement it in V, I am completely in favor of having as many things in V as possible, but it seems to me that `bfloat16` is something of a very specific use to have it in vlib , that's why I think it's okay to think about doing it in VSL, at least at first

I agree with having posit in vlib tho

**Reply by dumblob** [reply]
> Depends. It would be good to see an implementation in V, so the performance and accuracy could be verified.
> If posits really are a "drop-in" replacement for regular floats, and really are faster, it might be a good idea to replace the floating point code in V altogether.

Posits are "faster" in that regard, that they much better spread the accuracy among the spectrum of representable numbers unlike binary floats which have the highest accuracy around `+-0` and lowest around `+-inf`. But because common operations are not around `+-0` but rather in `0.1 .. some_billions`, then posits offer a sligtly better trade off and one can use posits with lower precision and then one gains a slight performance benefit (not on all CPU archetictures, but anyway).

> However, other considerations include size differences in both V and user code, as well conversions... will the numbers look the same (or enough the same) such that converting code to V will work the same? If the answer is "no, you'll get different results from the same calculations", that's a negative effect - even if the new results are "more correct".

Well, posits are still approximate numbers, so there is probably no clear cut. If you want something with a clear cut, you'd have to choose [decimal floats](https://github.com/vlang/v/discussions/11855 ) (ideally ["bigdecimals"](https://www.bytereef.org/mpdecimal/ ) as they have truly comparable speed to fixed-size decimal floats).
